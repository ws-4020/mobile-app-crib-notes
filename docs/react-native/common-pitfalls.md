---
title: ハマりがちな落とし穴
sidebar_label: Overview
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
