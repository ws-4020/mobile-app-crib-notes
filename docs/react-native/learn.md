---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。
