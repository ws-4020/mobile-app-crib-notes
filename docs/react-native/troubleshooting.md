---
title: トラブルシュート
---

React Nativeで開発する中で起きた困り事や不具合についてまとめています。
