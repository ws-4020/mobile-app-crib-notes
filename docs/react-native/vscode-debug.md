---
title: Visual Studio Codeの設定
---

ここではVisual Studio Code (以下VS Code) での開発に際して有用なプラグインをご紹介します。

## React Native Tools

VS Codeプラグインである`React Native Tools`を利用することで開発中のモバイルアプリケーションをデバッグできます。

VS Codeで、React NativeのTypeScriptコードをデバッグできるように、[React Native Tools](https://marketplace.visualstudio.com/items?itemName=msjsdiag.vscode-react-native)をセットアップします。


### React Native Toolsのインストール

VS CodeのReact Nativeの拡張ツールである、`React Native Tools`プラグインをインストールします。

Visual Studioマーケットプレイスの[React Native Toolsページ](https://marketplace.visualstudio.com/items?itemName=msjsdiag.vscode-react-native)の`Install`をクリックしてインストール。

または、VS Codeの「拡張機能」で`React Native Tools`検索してインストール。

### 起動構成を追加する

プロジェクトルート配下の`/.vscode/launch.json`に以下を追加してください。ファイルが無い場合は新規に作成してください。

```json title="/.vscode/launch.json"
{
  "configurations": [
    {
      "name": "Attach to packager",
      "cwd": "${workspaceFolder}",
      "type": "reactnative",
      "request": "attach",
      "port": "19000"
    }
  ]
}
```

:::info
portにはMetro Bundlerのポート番号を指定してください。デフォルトはExpoの場合19000、Expoを使用しない場合は8081です。

ポート番号をデフォルトから変更している場合はその番号を指定してください。
:::

### デバッグ

#### 実機の場合

1. 任意の起動方法にてアプリ起動した後に端末をシェイクし、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

#### シミュレータの場合

##### iOSシミュレータ

1. 任意の起動方法にてアプリ起動した後に`Ctrl+cmd+z`を押下し、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

##### Androidシミュレータ

1. 任意の起動方法にてアプリ起動した後にWindows環境の場合`Ctrl+m`、 Mac環境の場合`cmd+m`を押下し、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

## Lint、フォーマッタ

- [Prettier - Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
- [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

## その他

- [GitLens — Git supercharged](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens)
- [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)
- [Path Intellisense](https://marketplace.visualstudio.com/items?itemName=christian-kohler.path-intellisense)
- [Color Highlight](https://marketplace.visualstudio.com/items?itemName=naumovs.color-highlight)
- [Rainbow Brackets](https://marketplace.visualstudio.com/items?itemName=2gua.rainbow-brackets)
- [Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one)
