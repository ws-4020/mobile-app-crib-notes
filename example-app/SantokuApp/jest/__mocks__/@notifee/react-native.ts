/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

export default mock;
