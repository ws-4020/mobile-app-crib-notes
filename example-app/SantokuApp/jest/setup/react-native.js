/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
