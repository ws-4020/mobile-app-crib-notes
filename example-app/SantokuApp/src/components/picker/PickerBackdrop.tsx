import {composePressableStyles} from 'framework/utilities';
import React, {useMemo} from 'react';
import {Modal as RNModal, ModalProps, Pressable, PressableProps, StyleSheet, ViewProps} from 'react-native';
import Reanimated, {useAnimatedStyle, WithTimingConfig} from 'react-native-reanimated';

import {usePickerBackdropUseCase} from './usePickerBackdropUseCase';

export const DEFAULT_COLOR = 'black';
export const DEFAULT_OPACITY = 0.4;
export const DEFAULT_FADE_IN_DURATION = 300;
export const DEFAULT_FADE_OUT_DURATION = 150;

type Props = Reanimated.AnimateProps<ViewProps> & {
  isVisible: boolean;
  onPress?: () => unknown;
  afterFadeIn?: (finished?: boolean) => unknown;
  afterFadeOut?: (finished?: boolean) => unknown;
  fadeInDuration?: number;
  fadeOutDuration?: number;
  opacity?: number;
  pressableProps?: PressableProps;
  modalProps?: ModalProps;
  fadeInConfig?: WithTimingConfig;
  fadeOutConfig?: WithTimingConfig;
};

export const PickerBackdrop: React.FC<Props> = ({
  isVisible,
  opacity = DEFAULT_OPACITY,
  onPress,
  afterFadeIn,
  afterFadeOut,
  fadeInDuration = DEFAULT_FADE_IN_DURATION,
  fadeOutDuration = DEFAULT_FADE_OUT_DURATION,
  pressableProps: {style: pressableStyle, ...pressableProps} = {},
  modalProps: {style: modalStyle, ...modalProps} = {},
  fadeInConfig,
  fadeOutConfig,
  style,
  children,
  ...animatedViewProps
}) => {
  const {isModalVisible, animatedOpacity} = usePickerBackdropUseCase({
    isVisible,
    opacity,
    afterFadeIn,
    afterFadeOut,
    fadeInDuration,
    fadeOutDuration,
    fadeInConfig,
    fadeOutConfig,
  });

  const animatedStyle = useAnimatedStyle(() => ({opacity: animatedOpacity.value}));
  const composedPressableStyles = useMemo(
    () => composePressableStyles([styles.pressable, pressableStyle]),
    [pressableStyle],
  );
  const composedBackdropStyle = useMemo(() => StyleSheet.flatten([styles.backdrop, style]), [style]);

  return !isModalVisible ? null : (
    <RNModal
      visible
      statusBarTranslucent
      animationType="none"
      transparent
      // 戻るボタンが押されたとき（onRequestClose）は、背景がタップされたときと同じ振る舞いになるようにしておく。
      onRequestClose={onPress}
      style={modalStyle}
      {...modalProps}>
      <Pressable onPress={onPress} style={composedPressableStyles} {...pressableProps}>
        <Reanimated.View style={[composedBackdropStyle, animatedStyle]} {...animatedViewProps} />
      </Pressable>
      {children}
    </RNModal>
  );
};

const styles = StyleSheet.create({
  pressable: {
    ...StyleSheet.absoluteFillObject,
    display: 'flex',
  },
  backdrop: {
    flex: 1,
    backgroundColor: DEFAULT_COLOR,
  },
});
