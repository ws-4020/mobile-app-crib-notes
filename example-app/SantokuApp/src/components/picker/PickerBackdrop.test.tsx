import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {ModalProps, PressableProps, ViewProps} from 'react-native';
import Reanimated from 'react-native-reanimated';

import {DEFAULT_FADE_IN_DURATION, DEFAULT_FADE_OUT_DURATION, DEFAULT_OPACITY, PickerBackdrop} from './PickerBackdrop';

// If advancing a timer changes the state of a component, the timer must be run within an act.
// However, since act is `Thenable`, ESLint will issue a warning if you do not do something like await.
// For convenience, disable the relevant rule in this file.
/* eslint-disable @typescript-eslint/no-floating-promises */

// To get react-native-reanimated v2 animations to work with timer mocks, you need to use the `modern` timer.
// By default, Jest v26 uses `legacy` timer. So if you use `withReanimatedTimer` that use the default timer, animation does not start.
// https://jestjs.io/ja/docs/timer-mocks
jest.useFakeTimers('modern');

// TODO: Jest v27にアップデートできたら、withReanimatedTimerでテストを実装できるか検証する。
//       （JestのバージョンはExpoに依存しているので、Expoでのアップデートを待っている状態）
const startAnimation = () => act(() => jest.advanceTimersByTime(1));

describe('PickerBackdrop only with required props', () => {
  it('returns null if not visible', () => {
    const sut = render(<PickerBackdrop isVisible={false} />);
    // PickerBackdropがnullを返していることを確認したいがうまくやる方法が見当たらないので`toJSON`でnullになることを確認する。
    expect(sut.toJSON()).toBeNull();
  });

  it('renders successfully only with required props', () => {
    const sut = render(<PickerBackdrop isVisible testID="backdropAnimated" />);
    const animatedView = sut.getByTestId('backdropAnimated');
    //////////////////////////////////////////////////////////////////////////////////
    // 初期表示
    //////////////////////////////////////////////////////////////////////////////////
    // エラーが起きずにレンダリングされること
    expect(sut).toMatchSnapshot('Before animation started');
    expect(animatedView).toHaveAnimatedStyle({opacity: 0});

    startAnimation();

    // アニメーション中は`opacity`が変化すること
    act(() => jest.advanceTimersByTime(DEFAULT_FADE_IN_DURATION / 2));
    expect(animatedView).toHaveAnimatedStyle({opacity: DEFAULT_OPACITY / 2});
    expect(sut).toMatchSnapshot('Animating (fade in)');

    // アニメーションが完了すると`opacity`が設定値に到達していること
    act(() => jest.advanceTimersByTime(DEFAULT_FADE_IN_DURATION / 2));
    expect(animatedView).toHaveAnimatedStyle({opacity: DEFAULT_OPACITY});
    expect(sut).toMatchSnapshot('Just After fade in animation completed');

    // アニメーションが完了したあとは変化しないこと
    act(() => jest.advanceTimersByTime(10));
    expect(animatedView).toHaveAnimatedStyle({opacity: DEFAULT_OPACITY});
    expect(sut).toMatchSnapshot('Just After fade in animation completed');

    //////////////////////////////////////////////////////////////////////////////////
    // 非表示にする
    //////////////////////////////////////////////////////////////////////////////////
    sut.update(<PickerBackdrop isVisible={false} />);

    startAnimation();

    // アニメーション中は`opacity`が変化すること
    act(() => jest.advanceTimersByTime(DEFAULT_FADE_OUT_DURATION / 2));
    expect(animatedView).toHaveAnimatedStyle({opacity: DEFAULT_OPACITY / 2});
    expect(sut).toMatchSnapshot('Animating (fade out)');

    // アニメーションが完了するとコンポーネントが消えること
    act(() => jest.advanceTimersByTime(DEFAULT_FADE_OUT_DURATION / 2));
    expect(sut.container.children).toEqual([]);
    expect(sut).toMatchSnapshot('Just After fade out animation completed');

    // アニメーションが完了した後も少し時間を進めて、何も変わらないことを確認する
    act(() => jest.advanceTimersByTime(10));
    expect(sut.container.children).toEqual([]);
    expect(sut).toMatchSnapshot('Just After fade out animation completed');
  });
});

describe('PickerBackdrop with `onPress', () => {
  it('should be called on pressed', () => {
    const onPress = jest.fn();
    const sut = render(
      <PickerBackdrop
        isVisible
        onPress={onPress}
        testID="pickerBackDrop"
        pressableProps={{testID: 'pressable'}}
        modalProps={{testID: 'modal'}}
      />,
    );
    fireEvent.press(sut.getByTestId('pressable'));
    expect(onPress).toHaveBeenCalledTimes(1);
    fireEvent(sut.getByTestId('modal'), 'onRequestClose');
    expect(onPress).toHaveBeenCalledTimes(2);
  });
});
//
describe('PickerBackdrop with all props', () => {
  it('should be applied properly', () => {
    const onPress = jest.fn();
    const onLongPress = jest.fn();
    const afterFadeIn = jest.fn();
    /**
     * WithTimingConfigのeasingを取得できなかったため、以下のPropsは検証できていません。
     * - fadeIntConfig
     * - fadeOutConfig
     *
     * animatedPropsは取得できなかったため（Snapshot上にも存在していない）、検証できていません
     */
    const sut = render(
      <PickerBackdrop
        isVisible
        testID="animatedView"
        onPress={onPress}
        modalProps={{
          testID: 'modal',
          style: {backgroundColor: 'yellow'},
          statusBarTranslucent: false,
          animationType: 'fade',
          transparent: false,
          presentationStyle: 'fullScreen',
        }}
        pressableProps={{testID: 'pressable', style: {display: 'none'}, onLongPress}}
        animatedProps={{pointerEvents: 'none'}}
        style={{backgroundColor: 'green', borderColor: 'red'}}
        fadeInDuration={200}
        afterFadeIn={afterFadeIn}
      />,
    );
    expect(sut).toMatchSnapshot('PickerBackdrop with all props.');
    const modal = sut.getByTestId('modal');
    const pressable = sut.getByTestId('pressable');
    const animatedView = sut.getByTestId('animatedView');

    // assert modal
    fireEvent(modal, 'onRequestClose');
    const modalProps = modal.props as ModalProps;
    expect(modalProps.visible).toBe(true);
    expect(modalProps.statusBarTranslucent).toBe(false);
    expect(modalProps.animationType).toBe('fade');
    expect(modalProps.transparent).toBe(false);
    expect(modalProps.style).toEqual({backgroundColor: 'yellow'});
    expect(modalProps.presentationStyle).toBe('fullScreen');
    expect(onPress).toHaveBeenCalledTimes(1);

    // assert animatedView
    const animatedViewProps = animatedView.props as Reanimated.AnimateProps<ViewProps>;
    expect(animatedViewProps.style).toEqual({flex: 1, backgroundColor: 'green', borderColor: 'red', opacity: 0});

    // fadeInDurationで指定した時間の1msc前ではafterFadeInは実行されない
    act(() => jest.advanceTimersByTime(199));
    expect(afterFadeIn).not.toHaveBeenCalled();
    // fadeInDurationで指定した時間経過後は、afterFadeInが実行される
    act(() => jest.advanceTimersByTime(1));
    expect(afterFadeIn).toHaveBeenCalled();

    // assert pressable
    fireEvent.press(pressable);
    fireEvent(pressable, 'onLongPress');
    const pressableProps = pressable.props as PressableProps;
    expect(pressableProps.style).toEqual({bottom: 0, display: 'none', left: 0, position: 'absolute', right: 0, top: 0});
    expect(onPress).toHaveBeenCalledTimes(2);
    expect(onLongPress).toHaveBeenCalledTimes(1);

    //////////////////////////////////////////////////////////////////////////////////
    // 非表示にする
    //////////////////////////////////////////////////////////////////////////////////
    const afterFadeOut = jest.fn();
    sut.update(<PickerBackdrop isVisible={false} afterFadeOut={afterFadeOut} fadeOutDuration={100} />);

    startAnimation();
    // fadeOutDurationで指定した時間の1msc前ではafterFadeOutは実行されない
    act(() => jest.advanceTimersByTime(99));
    expect(afterFadeOut).not.toHaveBeenCalled();
    // fadeOutDurationで指定した時間経過後は、afterFadeOutが実行される
    act(() => jest.advanceTimersByTime(1));
    expect(afterFadeOut).toHaveBeenCalled();
  });
});
