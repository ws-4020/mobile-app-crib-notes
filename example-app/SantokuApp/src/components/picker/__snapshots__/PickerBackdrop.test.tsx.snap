// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerBackdrop with all props should be applied properly: PickerBackdrop with all props. 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  presentationStyle="fullScreen"
  statusBarTranslucent={false}
  style={
    Object {
      "backgroundColor": "yellow",
    }
  }
  testID="modal"
  transparent={false}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "display": "none",
        },
      ]
    }
    testID="pressable"
  >
    <View
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      entering={
        ZoomIn {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 500,
          "randomizeDelay": false,
        }
      }
      exiting={
        ZoomOut {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 300,
          "randomizeDelay": false,
        }
      }
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "red",
          "flex": 1,
        }
      }
      testID="animatedView"
    />
  </View>
</Modal>
`;
